design/anoncreds.md

```diff
@@ -0,0 +1,507 @@
+1. Try to find the interval (in `oldKeyTrustTime` array) the `issuanceTime` belongs to.
+    * If it's greater than the most right interval, then return the current value.
+    * If it belongs to an interval, then get the left value (`from`) of the interval.
+    * If it's in between intervals, then get the right interval, and get the left value (`from`)
+of this interval.
+1. Use generic logic to get the root of the State trie at the time `to` found above. 
+1. Lookup State Trie with the found root to find the state at that time (the same way as in Steps 1 and 2)
+
+So, we will have from 2 to 5 lookups for each request. 
+
+Result for the Example above:
+* `issuanceTime < A` => [A,B] => state at timeA => key1 (OK)
+* `A <= issuanceTime <= B` => [A,B] => state at timeA => key1 (OK)
+* `B < issuanceTime < C` => [C,D] => state at timeC => key2 (deprecated credentials)
+* `C <= issuanceTime <= D` => [C,D] => state at timeC => key2 (OK)
+* `D < issuanceTime < E` => [E,...] => state at timeE => key3 (deprecated credentials)
+* `issuanceTime > E` => [E,...] => state at timeE => key3 (OK)
+
+
+### REVOC_REG_DEF
+
+The Definition of revocation registry for the given CredDef.
+It contains public keys, maximum number of credentials the registry may contain,
+reference to the CredDef, plus some revocation registry specific data.
+
+#### REVOC_REG_DEF txn
+```
+{
+    "data": {
+        "uuid":"ZXzcdDLhCpGCYRHW82kjHd",
+        "type":"CL_ACCUM",
+        "credDefRef":"GEzcdDLhCpGCYRHW82kjHd",
+        "publicKeys": {},
+        "maxCredNum": 1000000,
+        "metadata": {
+            "tailsHash": "<SHA256 hash>",
+            "tailsLocation": "<URL>"
+        }
+        "oldKeyTrustTime": {                    (optional)
+            "from": "10", 
+            "to": "30",
+        }
+    },
+    
+    "reqMetadata": {
+        "issuerDid":"MMAD5g65TDQr1PPHHRoiGf",
+        .....
+    },
+    
+....
+}
+```
+* `uuid` is RevocRegDef's UUID. It's different from `issuerDid`.
+* `oldKeyTrustTime` can be set each time the accumulator key is rotated and defines
+ `the interval when we still can trust the previous value of the key`. 
+It is needed to deprecate credentials issued during the time when we suspect
+the keys were stolen.
+We can not always use revocation to deprecate old credentials, since revocation keys can
+be stolen as well.  
+ 
+
+#### Restrictions
+
+* Existing RevocRegDef (identified by the RevocRegDef `uuid`) can be modified/changed/evolved.
+That is rotation of keys is supported.
+* Only the `issuerDid` who created the RevocRegDef can modify it (that is we need to keep the ownership). 
+* It's not possible to create multiple entities with the same `uuid` (so, it's unique within the ledger).
+
+#### State
+
+We need to have two records for RevocRegDef in State Trie in order to have
+1. Simple referencing of RevocRegDef in the protocol (by RevocRegDef DID)
+1. Requirements 8
+
+Record 1:
```

Comment:
Again, dont need 2 keys, just one key holding the information needed to be given to the client is enough, that key can be looked up in cache by several filters

Reply:
+1